docs/proposals/promotion-strategy.md

```diff
@@ -0,0 +1,453 @@
+  commitStatuses:
+   - key: image-allowed
+     phase: pending
+   - key: code-freeze
+     phase: pending
+   - key: promotion
+     phase: pending
+```
+
+
+#### RevertCommit
+
+##### Example
+
+```yaml
+apiVersion: promoter.argoproj.io/v1alpha1
+kind: RevertCommit
+metadata:
+  labels:
+    promoter.argoproj.io/active-branch: dev
+spec:
+  repoRef:
+    name:
+  drySHA:
+  activeBranch: dev
+status:
+```
+
+#### CommitStatus
+
+A CommitStatus represents some piece of information about the state of a particular commit, for example "this commit is running in a cluster and is healthy" or "this commit is approved for promotion to production."
+
+Suppose the `environments/qal` branch holds the manifests for Argo CD Applications `example-qal-usw2` and `example-qal-use2`. Argo CD may create a CommitStatus indicating that hydrated commit cd5ba4a is `failed` because `example-qal-usw2` is synced and in a Degraded state.
+
+The controller responsible for promoting a change through environments could create a CommitStatus indicating that dry commit 82bf3ac is `successful` because the associated hydrated commits have been committed to all environments and have all been deployed and marked as healthy.
+
+The CommitStatus controller will synchronize the cluster state of a CommitStatus with the appropriate SCM (e.g. GitHub or Gitlab) via their commit APIs.
+
+SCM support must be implemented via a plugin interface so that the community can contribute support for their SCMs of choice.
+
+##### Relationship to other controllers
+
+This CRD will be created by other controllers, an example being an Argo CD specific Health/Sync status controller which will stay updated in realtime for Health/Sync status of an App (or group of apps). There is another type of Pre-Commit type check that can be generated such as kustomzie build type checks these types share the same CommitStatus CRD where it is used determines its type. If it's on a ProposedCommit, it's a pre commit check. If it's just on the PromotionStrategy activeCommitStatuses, it's an active health check used for promotion.
+
+##### Relationship to PromotionStrategy
+
+A CommitStatus by has no direct relationship to any PromotionStrategy (described below). To have any effect on promotion, the CommitStatus must be referenced (via a key selector) by a PromotionStrategy or a ProposedCommit which, in turn, is referenced by a PromotionStrategy. CommitStatus is meant to simply be an abstraction of the SCMs' commit status concepts. On its own, this abstraction should have no awareness of change promotion, GitOps, etc.
+
+##### Example
+
+```yaml
+apiVersion: commit-status.argoproj.io/v1alpha1
+kind: CommitStatus
+metadata:
+  labels:
+    commit-status.argoproj.io/key: argocd-health
+spec:
+  repoRef:
+    name:
+  # These fields roughly correspond to the fields available in the GitHub Checks API and the 
+  # Gitlab Commit "pipeline status" API. 
+  sha: xyz321  # Corresponds to the hydrated SHA
```

Comment:
Shouldn't this be a label as well to help the selection ?

Reply:
Possibly, but I think it might change quite often. Custom informer index on the user side might be the better option.

I've added a section explaining the mutability of this and the `repoRef` field.